pr.coverage function

mikapfl opened the pull request:
# Pull request

Please confirm that this pull request has done the following:

- [x] Tests added
- [x] Documentation added (where applicable)
- [x] Description in ``CHANGELOG.rst`` added

## Description

Add a function `pr.coverage` for Datasets and arrays with the functionality of pyCPA's `tools.dataAnalysis.check_coverage`. Closes: #7

JGuetschow commented on primap2/_overview.py:
```diff
@@ -0,0 +1,98 @@
+"""Summarization and descriptive statistics functions to get an overview of a data
```
A nice feature would be to be able to treat the time coordinate differently such that you have the option to have just 1 or 0 from the time coordinate. I  might want to know e.g. how many time series are present for a sector. In this case counting the individual time points is not helpful and selecting an individual time point doesn't give the full picture as there might be nan values for that year despite data being available for other time points.

mikapfl replied:
This would be functionality similar to `pr.sum_skip_all_na()` (in spirit), right? So, where `sum_skip_all_na` gives the option to sum along some dimension skipping all timeseries that are completely missing (all NA), you propose for the coverage() function an option to only count those timeseries that have at least one value, i.e. those that would not be skipped by `sum_skip_all_na`, right?

I have the feeling that there might be a more fundamental aggregation function like `sum_skip_all_na`, something like a `count_skip_all_na` or so, which would give this functionality also in other contexts. I'll think a bit about it.

JGuetschow replied:
So it could already be done by first collapsing the time dimensions using sum_skip_all_na and then running pr.coverage on the result. That's fine with me. It's good not to overload functions if it can already be done. But we have to take care to write a good documentation if users have to run several functions inn a row to achieve certain things (examples maybe).

mikapfl replied:
No, it is not yet possible, because the aggregation function we need is `all`, not `sum`. I will experiment with a new `all_skip_all_na` function (urgh, the name is consistent with `sum_skip_all_na`, but not very pretty) and see if it leads to a nice API.

JGuetschow reviewed (commented):
I found a bug and have one suggestion for improvements

mikapfl commented on primap2/_overview.py:
```diff
@@ -0,0 +1,98 @@
+        )
+
+
+class DatasetOverviewAccessor(_accessor_base.BaseDatasetAccessor):
```
The short names "cat" and "scen" are an implementation detail in the attrs, their public short names are "category" and "scenario" (i.e. without the terminology).

The behaviour of coverage() here is consistent with the behaviour of the `pr.loc[]` and `pr[]` getitem functions and I'd like to keep it consistent; however, it would of course be possible to add short aliases `scen` and `cat` everywhere and make them part of the public API. I think it could save some typing, but it can also be a bit non-obvious that `ds.pr["cat"] == ds.pr["category"] == ds["category (IPCC2006)"]`.

mikapfl replied:
And the second thing is a better error message when a non-existing dimension is used. I'll add that.

JGuetschow replied:
OK, then I misunderstood how the attrs are used. I think it's a bit confusing (at least to me as proven by example) that there are three names for one coordinate and one that is visible in the data structure in attrs can not be used for data access while a non-visible one can be used. Can you explain why it's  `scen` and `cat` in attrs if that can not be used in `.loc` statements?

mikapfl replied:
Nope, I can't explain it. I concur that it is confusing. It is not much work to add `scen` and `cat` aliases for `.loc` statements (and everywhere else), so I'll do that.

JGuetschow reviewed (approved):
All looks fine to me

mikapfl commented:
Ok, apart from fighting with windows tests and the readthedocs building (I guess it is good that we run the code in so many different environments in our CI, so we don't find these problems later on when using the software, but still...) I also did some changes:
* The aliasing now understands `scen` and `cat` everywhere.
* The aliasing is now more robust even when metadata gets lost (happens for some xarray functions when xarray isn't sure if metadata will be valid after the operation and `keep_attrs=True` is not set)
* I generalized the `coverage` function to N dimensions (with N>1), that was actually rather easy and might come in handy.

mikapfl commented:
Ah, and the problem with computing the coverage along the `entity` for entities with different dimensions actually turned out not to be relevant after all. It is well-defined when operating in the right order.

I also added a possibility to check if there are *any* values along a specific dimension (your example was timeseries), and count them. I changed `coverage` so it accepts boolean array. Using the building blocks of xarray, you can then simply do (e.g.):
```python
ds.notnull().any("time").pr.coverage("area", "entity")
```
This way it is more flexible than adding a specific option to the coverage function. For example, if you want to only count full time series (i.e. without any NaNs), you can do:
```python
ds.notnull().all("time").pr.coverage("area", "entity")
```
etc.
